EventedPLEG set container readiness to false when container was terminated

pandeykushagra51 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Immediately updates container readiness to false when container died when EventedPLEG feature is enabled.

#### Which issue(s) this PR is related to:
https://github.com/kubernetes/kubernetes/issues/128820
https://github.com/kubernetes/kubernetes/issues/126546
https://github.com/kubernetes/kubernetes/issues/106896
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed container readiness to update immediately after container termination when the EventedPLEG alpha feature is enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pandeykushagra51 committed:
kubelet: set container readiness to false when container terminates

Add readiness update in PLEG ContainerDied event handler to ensure containers marked for termination are immediately removed from service endpoints. This prevents the dangerous scenario where unhealthy containers continue receiving traffic during termination.

Add e2e test to verify container readiness behavior during pod termination and liveness probe failures.

Fixes issue 128820

toVersus commented:
/ok-to-test

pandeykushagra51 committed:
test/e2e/node: fix container readiness test during pod termination

- Remove unnecessary readiness/liveness probes to match real scenario
- Use alpine:latest with proper signal handling
- Update test logic to detect 1/2 containers ready during termination
- Replace deprecated wait.PollImmediate with wait.PollUntilContextTimeout
- Add proper error handling for pod deletion during monitoring
- Reduce test timeouts for faster execution

pandeykushagra51 committed:
test/e2e/node: fix linting errors in container readiness test

- Replace gomega.BeTrue() with gomega.BeTrueBecause() for better failure messages
- Fix forbidigo linting rule violations

pandeykushagra51 committed:
test/e2e/node: fixed import alias for k8s.io/apimachinery/pkg/api/errors to use apierrors

haircommander commented:
this makes sense to me. one nit but otherwise LGTM
/triage accepted
/priority important-soon

pandeykushagra51 committed:
chore: Update copyright year to 2025

pandeykushagra51 commented:
/retest

SergeyKanzhelev commented on test/e2e/node/container_readiness_during_termination.go:
```diff
@@ -0,0 +1,259 @@
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+var _ = SIGDescribe("Container Readiness During Termination", func() {
```
Add `NodeConformance` since the test doesn't require any special environment

pandeykushagra51 replied:
added NodeConformance

pandeykushagra51 replied:
done

SergeyKanzhelev commented on test/e2e/node/container_readiness_during_termination.go:
```diff
@@ -0,0 +1,259 @@
+			framework.Logf("Pod status during termination: %d/%d containers ready", readyContainers, len(pod.Status.ContainerStatuses))
+
+			// If container is being killed due to crash, readiness should be updated
+			if readyContainers < len(pod.Status.ContainerStatuses) {
```
so this test will succeed if both containers will be marked as NotReady right away? Should we expect 1 container being not ready?

SergeyKanzhelev commented on test/e2e/node/container_readiness_during_termination.go:
```diff
@@ -0,0 +1,259 @@
+						Lifecycle: &v1.Lifecycle{
+							PreStop: &v1.LifecycleHandler{
+								Exec: &v1.ExecAction{
+									Command: []string{"sh", "-c", "sleep 30"},
```
it will be WAY MORE reliable test to let it be something huge... like infinity for all practical puroses. Same as TerminationGracePeriodSeconds. 

Then your test will wait for one container becomeing NotReady and another keeps being Ready - terminating.

Then you can kill the Pod with force flag.

SergeyKanzhelev replied:
The way test implemented - it will likely flake at some point

pandeykushagra51 committed:
test/e2e/node: update container readiness test to address PR feedback

- Add NodeConformance decorator
- Use Eventually/Consistently for reliability
- Implement infinite timeouts for stability
- Remove redundant test case

pandeykushagra51 commented:
/retest

pandeykushagra51 commented:
> /hold
> 
> Lot's of changes needed in PR description:
> 
> 1. Update release notes. Since it is only fixed in case of Evented PLEG - note it in release notes. Also no need to tell that no action required. Just remove the whole "Action required" part.
> 2. Sine the PR only fixes Evented PLEG case - please do NOT use "Fixes" with the mentioned bugs. Those will NOT be fixed by this PR without the alpha feature enabled.
> 3. Why the section "What this PR does / why we need it:" is in code block? Make it a normal text.
> 
> Also my understanding is that the container is marked as not ready when Terminated, not while terminating. Retitling:
> 
> /retitle set container readiness to false when container was terminated
> 
> The termintING containers should keep running probes.

@SergeyKanzhelev thanks for flagging these, initially I was not aware that evented PLEG is an alpha features so I missed to mention that. I have updated PR description as suggested. Please take a look once you have some time.

pandeykushagra51 commented:
pinging @haircommander  @SergeyKanzhelev for review.
/assign @SergeyKanzhelev

pandeykushagra51 commented:
pinging @SergeyKanzhelev @haircommander for review

SergeyKanzhelev commented:
> /hold
> 
> Lot's of changes needed in PR description:
> 
> 1. Update release notes. Since it is only fixed in case of Evented PLEG - note it in release notes. Also no need to tell that no action required. Just remove the whole "Action required" part.
> 2. Sine the PR only fixes Evented PLEG case - please do NOT use "Fixes" with the mentioned bugs. Those will NOT be fixed by this PR without the alpha feature enabled.
> 3. Why the section "What this PR does / why we need it:" is in code block? Make it a normal text.
> 
> Also my understanding is that the container is marked as not ready when Terminated, not while terminating. Retitling:
> 
> /retitle set container readiness to false when container was terminated
> 
> The termintING containers should keep running probes.

You overdone it. You can still mention the bug, just do not prefix it with "Fixes".

Also release notes should not ask to enable EventedPLEG. It should say that WHEN EventedPLEG is enabled, the bug is fixed


/retitle EventedPLEG set container readiness to false when container was terminated

pandeykushagra51 committed:
test: add EventedPLEG feature gate requirement to container readiness test

pandeykushagra51 committed:
test: move EventedPLEG feature gate to SIGDescribe level

pandeykushagra51 commented:
> > /hold
> > Lot's of changes needed in PR description:
> > 
> > 1. Update release notes. Since it is only fixed in case of Evented PLEG - note it in release notes. Also no need to tell that no action required. Just remove the whole "Action required" part.
> > 2. Sine the PR only fixes Evented PLEG case - please do NOT use "Fixes" with the mentioned bugs. Those will NOT be fixed by this PR without the alpha feature enabled.
> > 3. Why the section "What this PR does / why we need it:" is in code block? Make it a normal text.
> > 
> > Also my understanding is that the container is marked as not ready when Terminated, not while terminating. Retitling:
> > /retitle set container readiness to false when container was terminated
> > The termintING containers should keep running probes.
> 
> You overdone it. You can still mention the bug, just do not prefix it with "Fixes".
> 
> Also release notes should not ask to enable EventedPLEG. It should say that WHEN EventedPLEG is enabled, the bug is fixed
> 
> /retitle EventedPLEG set container readiness to false when container was terminated

I’ve made the suggested changes. Please take a look whenever you have a moment.

pandeykushagra51 commented on pkg/kubelet/kubelet.go:
```diff
@@ -2492,6 +2492,8 @@ func (kl *Kubelet) syncLoopIteration(ctx context.Context, configCh <-chan kubety
 			if containerID, ok := e.Data.(string); ok {
+				parsedContainerID := kubecontainer.BuildContainerID(kl.containerRuntime.Type(), containerID)
+				kl.statusManager.SetContainerReadiness(logger, e.ID, parsedContainerID, false)
 				kl.cleanUpContainersInPod(e.ID, containerID)
```
@SergeyKanzhelev  please take a look
